### ISSUE TITLE
dmarc_reports_last_sent not created or updated

### ISSUE THREAD

**Issue by pdreissen** [rspamd/rspamd#2101]
### Classification (Please choose *one* option):

* [ ] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Serious bug
* [X] Other bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:
1.7.1
### Operation system, CPU, memory and environment:
Debian 9 / Nehalem-C 2 cores/ 8Gb
### Description (Please provide a descriptive summary of the issue):
When activating the dmarc reports they are not send because the file: $DBDIR/dmarc_reports_last_sent is not created. When touching this file reports are going to be sent however the file is not updated with the last sent value (timestamp ?)
### Steps to reproduce:
Create dmarc.conf in local.d with the following contents:

reporting = true;
send_reports = true;
report_settings {
  org_name = "companyname";
  domain = "domain.net";
  email = "postmaster@domain.net";
  helo = "rspamd.domain.net";
  retries = 5;
}

note: domain names in the config file have been changed.

### Expected results:
Reports being sent after 24hr +
### Actual results:
Reports weren't send, only after touching the file in the subject.
### Configuration:
reporting = true;
send_reports = true;
report_settings {
  org_name = "companyname";
  domain = "domain.net";
  email = "postmaster@domain.net";
  helo = "rspamd.domain.net";
  retries = 5;
}

**Comment by fatalbanana** [comment]
This file is updated merely on shutdown. If it doesn't exist on startup, reports are sent immediately (if data is available).

Issue is not clear to me I'm afraid.

**Comment by pdreissen** [comment]
Well the file isn't created at all, not @ shutdown not at startup. Reports are being sent out however only when rspamd is restarted. If rspamd runs for a couple of days no report is being sent out, when rspamd is restart then, reports start sending out immediately, but still no file is created!

**Comment by jangrewe** [comment]
Sign me up for this one. I once triggered a huge flush by touching that file (so at least my config works!), after seeing the error message in the log all the time, but since then there's silence...

**Comment by TDannhauer** [comment]
I have the same issuw with Rspamd 1.8.3 and Debian buster (Testing)

**Comment by rseichter** [comment]
Same issue with Rspamd 1.8.3 on Gentoo Linux. The state file `/var/lib/rspamd/dmarc_reports_last_sent` is never created, meaning there is no control over when DMARC reports are sent. Very annoying.

**Comment by github-k8n** [comment]
My IPv6 got temporarily blacklisted by google, probably due to the DMARC reports coming in more than once a day after a few restarts due to this issue..

**Comment by andryyy** [comment]
I doubt it.

**Comment by github-k8n** [comment]
Well, I didn't send much else to google... after disabling DMARC reports and waiting a few days, I was allowed again ... (glad that the IPv4 still worked as it was not used to connect to google earlier) MIGHT be coincidence .. however it was right after a bit more reports were sent so ... 
If you have ANY idea on how to find out WHY I was getting blocked by them, please let me know :) and no, postmaster tools are useless unless you ARE a spammer and you are sending bulk emails...

**Comment by PHPGangsta** [comment]
@github-k8n What was the exact reject message from Gmail? Were all mails to Gmail blocked with the same error message?
If you send mails to Gmail via IPv6, make sure your PTR is set up correctly, forward and reverse. That's a very common mistake why Gmail rejects mails via IPv6.

**Comment by rseichter** [comment]
What is the file `/var/lib/rspamd/dmarc_reports_last_sent` supposed to contain anyway? Is it in some textual format that would allow manually specifying a timestamp? I keep deleting entries in Redis in order to prevent reports from being sent whenever Rspamd is reloaded (happens a lot when experimenting with various config settings and rules).

I understand that the whole DMARC support is in an experimental stage, but the whole flying blind thing pushes me ever stronger back to using OpenDMARC instead.

**Comment by github-k8n** [comment]
> @github-k8n What was the exact reject message from Gmail? Were all mails to Gmail blocked with the same error message?
> If you send mails to Gmail via IPv6, make sure your PTR is set up correctly, forward and reverse. That's a very common mistake why Gmail rejects mails via IPv6.

Lol... Yes, I am aware of the basics. The server has correct PTR for both ipv6 and ipv4, it has SPF records, DKIM records, emails are correctly signed using dkim, it has correct DMARC headers etc, the zone is DNSSEC secured, TLSA (DANE) records, MTA-STS,. .. you name it..  Everything went just fine and dandy up until the day that rspamd sent too many dmarc reports out to google which lead to ALL emails being rejected from that IPv6 by google. (IPv4 still worked). Not sending anything from that IPv6 (i.e. forcing everything through IPv4 which was accepted without issue by google...) and after a few days google was again accepting emails.

I now disabled sending DMARC reports using
send_reports = false;
in /etc/rspamd/local.d/dmarc.conf
which is a shame but it seems that too many people set up dmarc records, specify an email address for reporting and then reject emails to that address (seen with bigger companies as well), so they don't seem to care about reporting that much .. thus I will not risk another issue with google or other servers because of that ... (Whether the issue WAS caused by rspamd sending too many emails or not is not entirely clear but for me the correlation points to a too high risk to continue sending them at all...)


@PHPGangsta if you are interested below is the answer from the google servers during the SMTP connection (My IPv6 redacted)

  mailauth-reports@google.com
    host aspmx.l.google.com [2a00:1450:400c:c08::1b]
    SMTP error from remote mail server after pipelined end of data:
    550-5.7.1 [XXXXXX      12] Our system has detected that this message
    550-5.7.1 is likely unsolicited mail. To reduce the amount of spam sent to
    550-5.7.1 Gmail, this message has been blocked. Please visit
    550-5.7.1  https://support.google.com/mail/?p=UnsolicitedMessageError
    550 5.7.1  for more information. w17si399235wro.239 - gsmtp


As you can see, absolutely useless...
Interestingly enough it seemed to wait for the message to be sent before rejecting it ...so it might not be a complete "blacklist" per se but maybe a "blacklist" in the spam software.. but ANY mail sent during those few days was rejected similarly .. even the most simple standard email with nothing else than text..

**Comment by jangrewe** [comment]
I eventually got it to work (months ago, so the details are a bit hazy) by an obscure combination of manually creating the lock file, deleting it again, restarting rspamd, sacrificing the blood of my (yet-to-be) first born, and everything not necessarily in that order.
The only thing i can say for sure is that it must have been around 7:30 PM, because that's when the reports are now being sent everyday, reliably.

Sorry for not being able to provide a clear solution, but at least there's hope!

**Comment by rseichter** [comment]
Happy Birthday, issue. ☹️

**Comment by heraklit256** [comment]
I can reproduce this issue with rspamd 1.9.4 .

@jangrewe: Would you mind posting your steps to make DMARC reporting working again here?

**Comment by jangrewe** [comment]
@heraklit256 After ~8 months, i'm afraid i can't. As mentioned it was some combination of manually creating the file and restarting rspamd, always keeping an eye on the log if there were any messages. And then it suddenly just worked.

**Comment by heraklit256** [comment]
@jangrewe: Thank you - something working all of a sudden is always unpleasant.

Anyway, I will try to fix this on my system without reinstalling rspamd and/or redis. Regarding the `dmarc_reports_last_sent` file (which is not present on my system), the [DMARC module documentation](https://rspamd.com/doc/modules/dmarc.html) suggests reporting information is stored in redis if `reporting = true;` is set.

**Comment by heraklit256** [comment]
Well, 24 hours later and after manually creating `dmarc_reports_last_sent` (and setting owner to `rspamd:rspamd`):
```
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1250: Couldnt extract reporting address: mailto:postmaster+dmarc@[REDACTED]
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1275: No reporting address for [REDACTED]
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1106: Deleted reports for [REDACTED]
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1250: Couldnt extract reporting address: mailto:dmarc-rua@[REDACTED]
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1275: No reporting address for [REDACTED]
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1106: Deleted reports for [REDACTED]
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1250: Couldnt extract reporting address: mailto:rua@fastmaildmarc.com
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1275: No reporting address for ftml.net
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1106: Deleted reports for ftml.net
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1250: Couldnt extract reporting address: mailto:mailauth-reports@google.com
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1275: No reporting address for gmail.com
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1106: Deleted reports for gmail.com
2019-08-15 10:42:05 #90288(controller) <arpe4b>; lua; dmarc.lua:1305: No more domains to generate reports for
```
Something is broken here. Will enable debugging for `dmarc` module and report back.

**Comment by jangrewe** [comment]
see? magic! :grin:

**Comment by heraklit256** [comment]
Yes. Actually, it seems like manually creating the file "solved" this issue. I think I will open another one as the log lines above indicate another bug.

Thanks for your advice.

**Comment by rseichter** [comment]
Has anybody already verified if this is still an issue with release 2.0 ?
